### DISCUSSION TITLE
Is there a way to rename attachments & their links when publishing?

### DISCUSSION THREAD

**Discussion by fr3fou** [ObsidianPublisher/obsidian-github-publisher#341] [Q&A — Plugin]
I have image attachments which have spaces in the name and I'd like to rename those to use a dash (and respectively rename the _links_ to them as well_). Is that possible? 
I know I can do the latter using a content regex replacer but I'm not sure if I can rename the attachments themselves.

```
![Pasted image 20231206204234.png|500](../../assets/Pasted image 20231206204234.png|500)
```

The reason why I need this is because Astro / other Markdown tools don't properly support attachments with spaces in them. Replacing the spaces with `%20` in order to URL encode them also doesn't work.

Was also wondering if there's a recommended way to remove the `|500` from the end. Regex replacement does seem like the most universal solution to a lot of stuff but it in my opinion it's clunky to use it for this.

**Comment by Mara-Li** [comment]
There are a lot of possible settings around path. You can override attachment path in the embed, but also change the slugify settings in the second tab (content or path I forgot the name)

**Reply by fr3fou** [reply]
I tried using the slug option but that didn't affect neither the link nor the filename 🤔

![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/1344906/2b2b32a8-04ad-4b15-8f06-841c0de93939)

**Reply by Mara-Li** [reply]
Lower juste lower the texte (for example `HELLO` become `hello`).
You can also try the regex replace for path or the override settings for attachments

**Reply by fr3fou** [reply]
I tried this as well and it still published the files with spaces :/ 

![image](https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/1344906/ff081e6f-ab5b-4184-b953-80d846c3d4f4)

**Reply by Mara-Li** [reply]
Did you try with the regex settings?

**Reply by fr3fou** [reply]
<img width="541" alt="image" src="https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/1344906/b6149e75-11ea-4191-9a9c-72b1890e5419">

I about to but I find the documentation for this really confusing – I'm not sure what I'd have to set it to

**Reply by Mara-Li** [reply]
Sorry. I rewrite multiple time the docs. English is not my native language and I write and update everything myself. I try to be the most understable as possible, but even in french people can't understand me... So in English? 🫡

It's difficult to make a documentation everyone can understand.

In the screen provided, in the first case you must put the part you want to replace. Here it's the space, so put a space.
In the second part, use a dashe for example.

Leave empty for force push, It's not relevant here.

You can also use the regex path remplacement as it does the same but for all links, including attachments and notes.

**Reply by fr3fou** [reply]
Thanks for the help and patience so far! 

Here's what I have so far:

<img width="578" alt="image" src="https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/1344906/f84b096c-9297-4de3-acba-c31f46493351">

Unfortunately, even this doesn't work 🤔

I also tried this:

<img width="592" alt="image" src="https://github.com/ObsidianPublisher/obsidian-github-publisher/assets/1344906/e05b972b-7e22-4e07-a573-34460dc9d328">

but this doesn't work either. I think this might be related to #338 🤔

**Reply by Mara-Li** [reply]
Oh, you should take the beta WITH BRAT yes :)

**Reply by fr3fou** [reply]
same behavior with the beta release :/

**Reply by Mara-Li** [reply]
Hmm, i think your settings (maybe fixed folder?) doesn't allow changing the path

**Reply by fr3fou** [reply]
hmm, I was using fixed folder but changed to property key and obsidian path, both didn't work 🤔.
here's my entire config:
```json
{
  "github": {
    "branch": "new",
    "automaticallyMergePR": false,
    "dryRun": {
      "enable": true,
      "folderName": "github-publisher"
    },
    "tokenPath": "%configDir%/plugins/%pluginID%/env",
    "api": {
      "tiersForApi": "Github Free/Pro/Team (default)",
      "hostname": ""
    },
    "workflow": {
      "commitMessage": "[PUBLISHER] Merge",
      "name": ""
    },
    "verifiedRepo": true
  },
  "upload": {
    "behavior": "fixed",
    "defaultName": "src/content/blog",
    "rootFolder": "",
    "yamlFolderKey": "",
    "frontmatterTitle": {
      "enable": false,
      "key": "slug"
    },
    "replaceTitle": [
      {
        "regex": "/\\\\W+/gi",
        "replacement": "-",
        "type": "title"
      }
    ],
    "replacePath": [],
    "autoclean": {
      "enable": false,
      "excluded": []
    },
    "folderNote": {
      "enable": false,
      "rename": "index.md",
      "addTitle": {
        "enable": false,
        "key": "title"
      }
    },
    "metadataExtractorPath": ""
  },
  "conversion": {
    "hardbreak": false,
    "dataview": true,
    "censorText": [
      {
        "entry": "/!\\[\\[(.*)\\]\\]/gi",
        "replace": "![$1](../../assets/$1)",
        "flags": "",
        "after": false
      },
      {
        "entry": "/banner: (.*)/gi",
        "replace": "banner: ../../assets/$1",
        "flags": "",
        "after": false
      }
    ],
    "tags": {
      "inline": false,
      "exclude": [],
      "fields": []
    },
    "links": {
      "internal": true,
      "unshared": false,
      "wiki": true,
      "slugify": "disable"
    }
  },
  "embed": {
    "attachments": true,
    "overrideAttachments": [
      {
        "path": " ",
        "destination": "-",
        "forcePush": false
      }
    ],
    "keySendFile": [
      "banner"
    ],
    "notes": false,
    "folder": "src/assets",
    "convertEmbedToLinks": "keep",
    "charConvert": "->",
    "unHandledObsidianExt": [],
    "sendSimpleLinks": true,
    "useObsidianFolder": false
  },
  "plugin": {
    "shareKey": "share",
    "excludedFolder": [
      "Journals"
    ],
    "copyLink": {
      "enable": true,
      "links": "https://...",
      "removePart": [],
      "transform": {
        "toUri": true,
        "slugify": "strict",
        "applyRegex": []
      }
    },
    "setFrontmatterKey": "Set"
  }
}
```

**Reply by Mara-Li** [reply]
Yeah, just noticed that ObsidianPath doesn't support overridenPath. So I make an update on my side.
For the global replace path, normally you should have an option between fileName and filePath.
But, you use "behavior: fixed" and it disable this!
